### ISSUE TITLE
Font sizes not accurate to WCAG guidelines

### ISSUE THREAD

**Issue by jrudmann** [getflourish/Sketch-Color-Contrast-Analyser#6]
First off, thanks for creating this plugin. The company I work for is embracing accessibility and our designers use Sketch as their primary design tool.

The issue I noticed was a discrepancy in the font size the plugin considers large scale text and the actual guidelines from the WCAG 2.0 AA. Below are a couple statements from the guidelines:
- "For example, for a 72 PPI image, an author would need to use approximately 19 pt and 24 pt font sizes in order to successfully present images of large-scale text to a user."
- "For many mainstream body text fonts, 14 and 18 point is roughly equivalent to 1.2 and 1.5 em or to 120% or 150% of the default size for body text", which is typically 16px.

Therefore:
- 16px x 120% = 19px
- 16px x 150% = 24px.

If I'm working at 1x in Sketch and measure 14 sized text in bold or 19 sized text in normal weight the plugin considers that a pass if the contrast is 3:1. I compared Sketch text sizing with Photoshop pixel sizing and they appear to be identical. I believe in this case the large scale text threshold should actually be 19 sized text in bold and 24 sized text in normal.

Here is another article trying to better explain how point sizes translate to pixels: http://mcdpartners.com/lab/meeting-wcag-color-contrast-guideline/

**Comment by centerofme** [comment]
I was also just downloading this and realized a similar thing take for instance two colors I was just analyzing (one is text and one is a background color) ... the result which is 4.4:1 says it does not pass AA ...however in a different web based tool i got the following answer "The contrast ratio is: 4.40:1

Passed at Level AA for large text only: If the text is large text (at least 18 point or 14 point bold), the luminosity contrast ratio is sufficient for the chosen colours (#3D8900 and #ffffff)."  This being said, the tool is great and I'm sure it is difficult to write in those rules into there so I think when it comes to fail I'll double check another source to gain insight into how what might help fix it.  In this case my text was already larger than 14 and it was bold so it passes.

**Comment by florianpnn** [comment]
I have to admit it's a tricky one. I'm currently writing an improved version of this plugin and I came across the same problem.
Would the possibility to add your own "body copy" would be better? For example here you speak about web where the body copy is 16px, but what happens when you design for iOS where the body copy provided by Apple is 17pt (@1x)?

I'd be happy to hear your thoughts on that!

**Comment by marcopellegrini-vodeclic** [comment]
Hi your plugin is great!
There is more and more Designers like me and my team that works on a11y with Sketch, I really think your plugin have big growth margins.

I have a feature request for you: 
can you include in your plugin the sketch color palette in order to make the WCAG analysis?
Something that integrate these functions https://accessibility.oit.ncsu.edu/tools/color-contrast/accessible-color-palette.php?&colors=000000,ffffff,aeaeae,E2A156&main=aeaeae&level=AA

**Comment by Buvik** [comment]
Sketch operates with pixels, if I am not wrong. And as per WCAG 2.0: https://www.w3.org/TR/UNDERSTANDING-WCAG20/visual-audio-contrast-contrast.html

"The ratio between sizes in points and CSS pixels is 1pt = 1.333px, therefore 14pt and 18pt are equivalent to approximately 18.5px and 24px."

So **large text** in Sketch should be defined as above 24(pixels). It seems that it is now set to 18.

**Comment by getflourish** [comment]
Valid points. Indeed, one can not solely rely on the numbers. Fonts can behave very different when it comes to x-height and legibility. I like the idea of @florianpnn about “add your own body copy” or say, custom settings. This could also help dealing with various screen resolutions etc.

**Comment by barts** [comment]
I am looking at the code of the plugin and it's pretty clear that it's checking for things at 18px Regular or 14px Bold as its minimums, which is not correct. It should be using 24px Regular and 18.5px Bold (or more accurately 23.99px Regular and 18.66px Bold). 

Yes there are variations between fonts but you're already way off by 4-6px, and returning a lot of false positives for people looking at UI type (which might be as small as 11px to 16px, which would still need to pass the "Normal" size test even when they're bold).

**Comment by getflourish** [comment]
While you may be right that font sizes in Sketch might not be the exact same numeric pixel value in every environment, the values used by the plugin relate to points as states in the guidelines.

Sketch uses native NSFont to display text layers. NSFont uses points for text sizes. https://developer.apple.com/documentation/appkit/nsfont/1525977-init

I see that the plugin might need to communicate this clearly to users, but I don’t think that there is a wrong calculation in place.

**Comment by barts** [comment]
@getflourish But points aren't a fixed size. NSFont is using 72 ppi, where 14pt = 14px. But the WCAG guidelines are based on CSS3 points, where 1pt = 1.333px, so 14pt = 18.5px (roughly). When someone sets their type in Sketch to "14", it'll look the same when they set their font to "14px" in CSS. But the WCAG measurements in px are 24px Regular and 18.5px Bold, not 14px and 18px. 

So you shouldn't be using the NSFont sizes at face value, you should be compensating to match the WCAG's actual requirements for contrast (which need the fonts to be bigger than you're checking for).

**Comment by getflourish** [comment]
https://github.com/getflourish/Sketch-Color-Contrast-Analyser/issues/22
